cluster-autoscaler/core/static_autoscaler.go

```diff
@@ -719,7 +719,8 @@ func (a *StaticAutoscaler) obtainNodeLists(cp cloudprovider.CloudProvider) ([]*a
 	// Treat those nodes as unready until GPU actually becomes available and let
 	// our normal handling for booting up nodes deal with this.
 	// TODO: Remove this call when we handle dynamically provisioned resources.
-	allNodes, readyNodes = gpu.FilterOutNodesWithUnreadyGpus(cp.GPULabel(), allNodes, readyNodes)
+	allNodes, readyNodes = a.processors.CustomResourceProcessor.FilterOutNodesWithUnreadyResource(a.AutoscalingContext,
+		gpu.ResourceNvidiaGPU, cp.GPULabel(), allNodes, readyNodes)
```

Comment:
I don't particularly like this, it feels like doing the refactor half-way - I still couldn't use a processor to support "GPU hack" for my own custom resource, because it's only ever going to be called for GPU. I think it would be better to dump the gpu related parameters from the method signature, so it looks like this:
```
FilterOutNodesWithUnreadyResource(a.AutoscalingContext, allNodes, readyNodes)
```

This way it's the implementation of the processor that decides what resources need to be initialized to consider the node ready. The default implementation could than be renamed to GpuCustomResourceProcessor and use ResourceNvidiaGPU label as well as GPULabel() provided by cloudprovider.

---

I think the difference in our approaches is what logic should be customizable by processor - your PR goes into "how GPUs are handled direction", but I don't think there are actually many different ways to handle this. I'd much rather assume the processor implements the current "GPU hack" or something very similar and allow customizing "what resource are we doing the hack for / how do we recognize those resources". I could even imagine the implementation of the processor not looking at resource request, but something else entirely - ex. we use basically the same "GPU hack" to handle "ignore-taints" now. With the processor implemented the way I describe that logic could be re-implemented as a combination of NodeInfoProcessor and CustomResourceProcessor without any explicit support in core (note: I'm not asking you to reimplement it, just using it as example :) ).

NodeGroupSetProcessor is a similar example with mulitple implementations that use fundamentally the same logic and only differ in details (what labels they look at, etc).

Reply:
I took a long hard look at my implementation and I must agree that your approach would be superior here - it would be much more flexible and customisable. It will require a significant bit more refactoring of the context in which these processor methods are executed, but it is definitely doable.

Regarding `GetNodeTargetResource`, I think that the implementation returning something like a slice with triples `(resource, resourceType, resourceCount)`, eg. `("nvidia.com/gpu", "nvidia-tesla-k80", 2)` should be flexible enough to work in all current contexts - I would not want to "mix" our approaches into one PR, it would be better to follow yours strictly.